cypress/integration/x-msc2716-historical-import/historical-import.spec.ts

```diff
@@ -0,0 +1,531 @@
+        })
+            .then((res) => {
+                // Registration success
+                if (res.status === 200) {
+                    return;
+                }
+
+                const errcode = res.body.errcode;
+
+                // User already registered and good to go
+                if (res.status == 400 && errcode === "M_USER_IN_USE") {
+                    return;
+                }
+
+                const errorMessage = res.body.error;
+                throw new Error(`ensureVirtualUserRegistered failed to register: (${errcode}) ${errorMessage}`);
+            });
+    });
+}
+
+function setupRoomWithHistoricalMessagesAndMarker({
+    synapse,
+    asMatrixClient,
+    virtualUserIDs,
+}: {
+    synapse: SynapseInstance;
+    asMatrixClient: MatrixClient;
+    virtualUserIDs: string[];
+}) {
+    // As the application service, create the room so it is the room creator
+    // and proper power_levels to send MSC2716 events. Then join the logged
+    // in user to the room.
+    cy.wrap(null).then(async function() {
+        const resp = await asMatrixClient.createRoom({
+            // FIXME: I can't use Preset.PublicChat because Cypress doesn't
+            // understand Typescript to import it
+            preset: "public_chat" as Preset,
+            name: "test-msc2716",
+            room_version: "org.matrix.msc2716v3",
+        });
+        cy.wrap(resp.room_id) .as('roomId');
+    });
+
+    cy.wrap(null).then(function() {
+        // Join the logged in user to the room
+        cy.joinRoom(this.roomId);
+
+        // Then visit the room
+        cy.visit("/#/room/" + this.roomId);
+    });
+
+    // Send 3 live messages as the application service.
+    // Then make sure they are visible from the perspective of the logged in user.
+    cy.wrap(null).then(async function() {
+        // Send 3 messages and wait for them to be sent
+        const liveMessageEventIds = [];
+        for (let i = 0; i < 3; i++) {
+            // Send the messages sequentially waiting for each request
+            // to finish before we move onto the next so we don't end up
+            // with a pile of live messages at the same depth. This
+            // gives more of an expected order at the end.
+            const { event_id: eventId } = await asMatrixClient.sendMessage(this.roomId, null, {
+                body: `live_event${i}`,
+                msgtype: "m.text",
+            });
+            liveMessageEventIds.push(eventId);
+        }
+
+        // Make this available for later chains
+        cy.wrap(liveMessageEventIds).as('liveMessageEventIds');
+
+        // Wait for the messages to show up for the logged in user
+        waitForEventIdsInClient(liveMessageEventIds);
+
+        cy.wrap(liveMessageEventIds).as('liveMessageEventIds');
+    });
+
+    cy.wrap(null).then(function() {
+        // Make sure the right thing was yielded
+        expect(this.liveMessageEventIds).to.have.lengthOf(3);
```

Comment:
I have a bunch of `@typescript-eslint/no-invalid-this` (`Unexpected 'this`) errors across these tests because I am referencing multiple aliased values per enqueued callback.

[`cy.get(...)`](https://docs.cypress.io/api/commands/get#Syntax) does not support grabbing multiple items and using it to grab every value we need would end up in callback hell, related https://stackoverflow.com/questions/55298710/how-to-get-values-of-multiple-aliases-without-introducing-callback-hell-in-cypre

I see in other tests, that we assign to locally scoped variables but that doesn't seem very Cypress'y and slightly error prone if you get the queuing wrong. Which can lead to easy `undefined` footguns with you only catching it because of the symptom downstream.

It would be a lot nicer to use something like `cy.all(...)` that dwelle crafted in https://stackoverflow.com/a/55313313/796832 and https://github.com/cypress-io/cypress/issues/915#issuecomment-568037175

TypeScript versions:

 - https://github.com/cypress-io/cypress/issues/915#issuecomment-1101631467
 - https://github.com/cypress-io/cypress/issues/915#issuecomment-901911155

Ex.
```js
it("test", () => {
    cy.wrap('one').as('one');
    cy.wrap('two').as('two');
    cy.wrap('three').as('three');

    cy.all([cy.get<number>("@one"), cy.get<string>("@two"), cy.wrap(true)]).then(([a, b, c]) => {
        console.log(a, b, c);
    });
});
```

Are we open to adding this?

Reply:
Asked in the [#element-dev room](https://matrix.to/#/!bEWtlqtDwCLFIAKAcv:matrix.org/$17ufvjvtm9yvfxvFmZ8WNSQmdTnWIOryGIFAj8zdYis?via=matrix.org&via=element.io&via=mozilla.org), @t3chguy already added this in one of his PR's: https://github.com/matrix-org/matrix-react-sdk/pull/8645/files#diff-845f38c4fbfce8fadc06d79d0bb879e309353eeaf6b1424622e4aafcf2cb35b6R50

We can copy that into this PR for now ⏩